keps/sig-storage/20200127-recover-from-resize.md

```diff
@@ -0,0 +1,153 @@
+
+Problem in allowing users to reduce size is - a malicious user can use this feature to abuse the quota system. They can do so by rapidly expanding and then shrinking the PVC. In general - both CSI and in-tree volume plugins have been
+designed to never perform actual shrink on underlying volume, but it can fool the quota system in believing the user is using less storage than he/she actually is. To solve this problem - we propose that although users are allowed to reduce size of their PVC (as long as requested size >= `pvc.Status.Capacity`), quota calculation will only use previously requested higher value. This is covered in more detail in [Implementation](#implementation).
+
+
+
+### Goals
+
+- Allow users to cancel previously issued volume expansion requests, assuming they are not yet successful or have failed.
+- Allow users to retry expansion request with smaller value than original requested size in `pvc.Spec.Resources`, assuming previous requests are not yet successful or have failed.
+
+### Non-Goals
+
+- As part of this KEP we do not intend to actually support shrinking of underlying volume.
+
+## Proposal
+
+As part of this proposal, we are mainly proposing three changes:
+
+1. Relax API validation on PVC update so as reducing `pvc.Spec.Resoures` is allowed, as long requested size is `>=pvc.Status.Capacity`.
+2. Add a new field in PVC api called - `pvc.Spec.AllocatedResources` for purpose of tracking used storage size. This field is only allowed to increase.
+3. Update quota code to use `max(pvc.Spec.Resources, pvc.Spec.AllocatedResources)` when evaluating usage for PVC.
+
+### Implementation
+
+We propose that by relaxing validation on PVC update to allow users to reduce `pvc.Spec.Resources`, it becomes possible to cancel previously issued expansion requests or retry expansion with a lower value if previous request has not been attemptd. In general - we know that volume plugins are designed to never peform actual shrinking of the volume, for both in-tree and CSI volumes. Moreover if a previously issued expansion has been successful and user
+reduces the PVC request size, for both CSI and in-tree plugins they are designed to return a successful response with NO-OP. So, reducing requested size will be a safe operation and will never result in data loss or actual shrinking of volume.
+
+We however do have a problem with quota calculation because if a previously issued expansion is successful but is not recorded(or partially recorded) in api-server and user reduces requested size of the PVC, then quota controller will assume it as actual shrinking of volume and reduce used storage size by the user(incorrectly). Since we know actual size of the volume only after performing expansion(either on node or controller), allowing quota to be reduced on PVC size reduction will allow an user to abuse the quota system.
```

Comment:
why does the "quota controller assume an actual shrinking of volume" if that's never even possible, couldn't that simply be changed to be "volume size can never be smaller than previous volume size". My hunch is that the quota controller is not actually "assuming" anything and that this statement is just an inference and not a representation of what the controller is actually doing (based on the other statements)

Reply:
There's a few problems, and I think it would be good to describe them in this kep, and why we can't solve them.

1) There's a race where volume expansion succeeds, but we failed to record the new size, and user requests a smaller volume in the meantime
2)  Most volume plugins have wrongly implemented the Expand call, and are not returning the actual size detected
3)  Kubernetes doesn't use the actual size returned by the plugin

Reply:
Yeah I did not include all the cases we discussed in Google doc that documents, why we can't solve of them. I will update the KEP.

Reply:
I have updated the failure conditions in alternatives considered.